File: 2017/5/src/main/java/com/mikeseagull/adventofcode2017/Part1.java
package com.mikeseagull.adventofcode2017;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Part1 {
    protected List<Integer> readFile(String fileName) {
        List<Integer> instructions = new ArrayList<Integer>();

        File inputFile = new File(this.getClass().getClassLoader().getResource(fileName).getFile());
        try (Scanner scanner = new Scanner(inputFile)) {
            while (scanner.hasNextLine()) {
                String line = scanner.nextLine();
                instructions.add(Integer.parseInt(line.trim()));
            }
            scanner.close();
        } catch (IOException e) {}
        return instructions;
    }
    public int main() {
        List<Integer> instructions = readFile("part1_input");
        int tracker = 0;
        int step_counter = 0;
        while (tracker < instructions.size()) {
            step_counter++;
            int original_tracker = tracker;
            int instruction = instructions.get(tracker);
            tracker += instruction;
            instructions.set(original_tracker, instruction + 1);
        }
        return step_counter;
    }
}

File: 2017/2/src/main/kotlin/com/mikeseagull/adventofcode2017/Part1.kt
package com.mikeseagull.adventofcode2017

open class Part1 {

    fun readFile(fileName: String): ArrayList<String> {
        var inputStream = this.javaClass::class.java.getResource(fileName).openStream()
        val lineList = mutableListOf<String>()
        inputStream.bufferedReader().useLines {
            lines -> lines.forEach {
            lineList.add(it)
        }
        }
        var lines = ArrayList<String>()
        lineList.forEach {
            lines.add(it)
        }
        return lines
    }
    fun parseLine(line: String): ArrayList<Int> {
        var numbers = ArrayList<Int>()
        var beginIndex = 0
        var index = line.indexOf("\t")
        while (index >= 0) {
            numbers.add(Integer.parseInt(line.substring(beginIndex, index)))
            //println(index)
            beginIndex = index + 1
            if (line.indexOf("\t", index + 1) < 0) {
                break
            } else {
                index = line.indexOf("\t", index + 1)
            }
        }
        numbers.add(Integer.parseInt(line.substring(index + 1)))
        return numbers
    }
    fun differenceOfMaxAndMin(numbers: ArrayList<Int>): Int {
        numbers.sortDescending()
        var max: Int = numbers.get(0)
        var min: Int = numbers.last()
        return max - min
    }
    open fun main(): Int {
        var lines = readFile("/part1_input")
        var checksum = 0
        for (line in lines) {
            var numbers = parseLine(line)
            var difference = differenceOfMaxAndMin(numbers)
            checksum = checksum + difference
        }
        return checksum
    }
}